clientv3/credentials/credentials.go

```diff
@@ -115,15 +139,33 @@ func (tc *transportCredential) Info() grpccredentials.ProtocolInfo {
 }
 
 func (tc *transportCredential) Clone() grpccredentials.TransportCredentials {
+	copy := map[string]string{}
+	tc.mu.Lock()
+	for k, v := range tc.addrToEndpoint {
+		copy[k] = v
+	}
+	tc.mu.Unlock()
 	return &transportCredential{
-		gtc: tc.gtc.Clone(),
+		gtc:            tc.gtc.Clone(),
+		addrToEndpoint: copy,
 	}
 }
 
 func (tc *transportCredential) OverrideServerName(serverNameOverride string) error {
 	return tc.gtc.OverrideServerName(serverNameOverride)
 }
 
+func (tc *transportCredential) Dialer(ctx context.Context, dialEp string) (net.Conn, error) {
```

Comment:
So, gRPC first calls `Dialer` and by the time it calls, `ClientHandshake`, the remote address will always have been populated in `addrToEndpoint`, right?

Reply:
Right, here's the dial call: https://github.com/grpc/grpc-go/blob/45bd2846a34b039c5f1e69b7202f118687156b34/internal/transport/http2_client.go#L167 and below it is the ClientHandshake call: https://github.com/grpc/grpc-go/blob/45bd2846a34b039c5f1e69b7202f118687156b34/internal/transport/http2_client.go#L212

Reply:
And here's where the authority gets set: https://github.com/grpc/grpc-go/blob/230def769105c46b2e597578b23002a6ac159dde/clientconn.go#L1142.

We're basically just trying to get the authority set to an appropriate value for each endpoint instead of having a single value for the entire set of endpoints.